ipatests/pytest_ipa/integration/resolver.py

```diff
@@ -0,0 +1,316 @@
+            raise Exception('No resolver backups found for host {}'.format(
+                self.host))
+        self.check_state_expected()
+        self._set_state(self.backups.pop())
+        logger.info(
+            'Restored resolver state for host %s, number of saved states: %s',
+            self.host, len(self.backups)
+        )
+
+    def has_backups(self):
+        """Checks if stack of backups is not empty"""
+        return bool(self.backups)
+
+    def check_state_expected(self):
+        """Checks if resolver configuration has not changed.
+
+        Raises AssertionError if actual configuration has changed since last
+        call to any method of Resolver
+        """
+        assert self._get_state() == self.current_state, (
+            'Resolver state changed unexpectedly at host {}'.format(self.host))
+
+    def __enter__(self):
+        self.backup()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.restore()
+
+    def _set_state(self, state):
+        self._apply_state(state)
+        logger.info('Applying resolver state for host %s: %s', self.host, state)
+        self.current_state = state
+
+    @abc.abstractclassmethod
+    def is_our_resolver(cls, host):
+        """Checks if the class is appropriate for managing resolver on the host.
+        """
+
+    @abc.abstractmethod
+    def _make_state_from_args(self, nameservers, searchdomains):
+        """
+
+        :param nameservers: list of ip addresses of nameservers
+        :param searchdomains: list of searchdomain, can be an empty list
+        :return: internal state object specific to subclass implementaion
+        """
+
+    @abc.abstractmethod
+    def _get_state(self):
+        """Acquire actual host configuration.
+
+        :return: internal state object  specific to subclass implementaion
+        """
+
+    @abc.abstractmethod
+    def _apply_state(self, state):
+        """Apply configuration to host.
+
+        :param state: internal state object  specific to subclass implementaion
+        """
+
+
+class ResolvedResolver(Resolver):
+    RESOLVED_RESOLV_CONF = {
+        "/run/systemd/resolve/stub-resolv.conf",
+        "/run/systemd/resolve/resolv.conf",
+        "/lib/systemd/resolv.conf",
+        "/usr/lib/systemd/resolv.conf",
+    }
+    RESOLVED_CONF_FILE = (
+        '/etc/systemd/resolved.conf.d/zzzz-ipatests-nameservers.conf')
+    RESOLVED_CONF = textwrap.dedent('''
+        # generated by IPA tests
+        [Resolve]
+        DNS={nameservers}
+        Domains=~. {searchdomains}
+    ''')
+
+    @classmethod
+    def is_our_resolver(cls, host):
+        res = host.run_command(
+            ['stat', '--format', '%F', paths.RESOLV_CONF])
+        filetype = res.stdout_text.strip()
+        if filetype == 'symbolic link':
+            res = host.run_command(['realpath', paths.RESOLV_CONF])
+            return (res.stdout_text.strip() in cls.RESOLVED_RESOLV_CONF)
+        return False
```

Comment:
I would rather suggest using `resolvectl status` and check that it returned exit code 0.

Reply:
The logic for detecting if resolv.conf is managed by systemd-resolved is "borrowed"  from https://github.com/freeipa/freeipa/blob/3fc75b15faea63c2468f70411e5604cc62bb4dce/ipalib/install/dnsforwarders.py#L37

What I like in the current approach is that it checks if resolved is actually managed by resolved. `systemctl ...` and `resolvectl status`would only check if it should be, but it can be easily overridden by replacing symlink with a file. Such change can be intentional - i.e. to simulate NM-managed or unmanaged resolver on modern system

Reply:
Ok, that makes sense. We adopted the approach in `dnsforwarders.py` from systemd-resolved and NM -- they use that to detect each other too.

Reply:
Ok, works for me, too